addons/web_editor/static/src/js/editor/odoo-editor/test/utils.js

```diff
@@ -284,13 +284,18 @@ export async function testEditor(Editor = OdooEditor, spec, options = {}) {
         editor.testMode = true;
         if (selection) {
             await setTestSelection(selection);
-            editor._recordHistorySelection();
         } else {
             document.getSelection().removeAllRanges();
         }
 
         // we have to sanitize after having put the cursor
         sanitize(editor.editable);
+        editor.historyReset();
+        editor.historyStep();
```

Comment:
This should probably have a comment.

Reply:
True.

```js
// The changes made by `sanitize` are included in a step which could
// well be rolled back if what the step function does triggers it. This
// is not how the editor normally behaves. Reset the history now.
```

Reply:
IMO it's more about the technical detail that in the tests we have to call sanitize manually after having put the selection and that means we do it after initializing the history (which is not the case when using toSanitize) but then we can undo the sanizite, which we shouldn't. What I mean is that the code is wrong without this change even when no rollback is triggered. It's just that the issue is invisible when there is no rollback, but it's there nonetheless.

Reply:
OK. I added a comment above `toSanitize: false`, one above the call to `sanitize`, and adapted this comment.

```js
// We disable the `toSanitize` option so we set the test selection on the
// raw, unsanitized HTML. We'll sanitize after having set the selection.
const editor = new Editor(testNode, Object.assign({ toSanitize: false }, options));
```

```js
// Now the selection is set we can finally sanitize.
sanitize(editor.editable);

// In normal circumstances the editor sanitizes its content in its
// constructor, before initializing the history. For the purposes of
// setting the test's selection, we only sanitize now, which means the
// changes made by `sanitize` are included in a step (which could well
// be rolled back if what the step function does triggers it), and the
// sanitization can be undone. This is not how the editor normally
// behaves. To make it closer to reality, we now reset the history.
editor.historyReset();
editor.historyStep();
```

Reply:
Better, thanks! Sorry to nitpick but I would take the opportunity to mention why we can't sanitize earlier, that is because the presence of the `[ ]` characters can affect the result of the sanitize, for example a node that is empty and thus should receive a BR will not receive one because it appears non empty due to its content being `[ ]` until we remove those and actually apply the selection.

Reply:
Actually we do this before initializing the editor:
```js
// Add the content to edit and remove the "[]" markers *before* initializing
// the editor as otherwise those would genererate mutations the editor would
// consider and the tests would make no sense.
testNode.innerHTML = spec.contentBefore;
```
The reason we need to sanitize _after_ setting it is to make sure the selection is properly set.

For instance:

`<p><a>abc</a></p>` + `setSelection(a, 1)` = `<p><a>abc[]</a></p>`

But,

`<p><a>abc</a></p>` + `sanitize()` +  `setSelection(a, 1)`
= `<p>\ufeff<a>\ufeffabc\ufeff</a>\ufeff</p>` + `setSelection(a, 1)`
= `<p>\ufeff<a>\ufeff[]abc\ufeff</a>\ufeff</p>`

`abc[] != `[]abc`